### PR TITLE
GF Signup: Dropdown Interval: Implement mobile sticky plan type selector behavior

### PR DESCRIPTION
**Pull request by jeyip**
Related to https://github.com/Automattic/growth-foundations/issues/296 and Attempt 1 at sticky headers https://github.com/Automattic/wp-calypso/pull/85158

## Proposed Changes

* Introduces the a sticky dropdown for mobile based screens

## Testing Instructions

* Check out this branch
* Navigate to /start/plans?flags=onboarding/interval-dropdown ( don't forget the feature flag! )
* Verify that no changes are seen in tablet and desktop views
* Switch to a mobile view and verify that a sticky plan selector dropdown is displayed
* Verify that the dropdown updates mobile plan pricing as expected

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/plans-grid/components/sticky-container.tsx
Comment (jeyip) [inline client/my-sites/plans-grid/components/sticky-container.tsx:43]: TODO:

I was doing some learning in this file. Will remove all inline comments from this file before shipping 🙂

File: client/my-sites/plans-grid/components/plan-type-selector/types.ts
Comment (jeyip) [inline client/my-sites/plans-grid/components/plan-type-selector/types.ts:25]: TODO:

Remove passed prop `isStuck` if unnecessary

**Comment by jeyip** [comment]
TODOS to self:

- Fix 48px offset for comparison grid stick plans header
- There's been a lot of tweaks in this PR. Make one more pass for code quality

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [inline client/my-sites/plans-features-main/index.tsx]: **Note:**

@chriskmnds I know that you mentioned wanting to keep responsive desktop / tablet / mobile concerns to css media query calculations ( and avoid the use of `isMobile` ) https://github.com/Automattic/wp-calypso/pull/85158#discussion_r1424955463

I did a decent amount of tinkering, but I couldn't figure out a clear way to proceed with media query mixins for this specific case without adding a non-trivial amount of complexity. This is because the sticky offset we'd like to adjust depends on `masterbarHeight`, which is calculated at runtime by our javascript, and inaccessible for evaluation in CSS.

One of the first things I did was try to experiment with conditional padding or margins, but I couldn't find a stickied target that wouldn't affect existing layouts when padded. The rest of my experimentation went similarly 🤔 

I might be missing something obvious though since I've been staring at this code for quite a while 🤔 Totally open to suggestions.

Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: Thanks @jeyip . I had a brief look at the code in general and it looks pretty good. I have some comments for reflection, which i'll post later.

For this case, if we do end up with unreasonable complexity, then fine to use other means. Also, if the logic stays outside `plans-grid`, then we don't really care about how we define "mobile". The plans-grid package will likely have its own "small", "medium", "large" dimensions that will be determined differently (through own listener), but anything that consumes these media queries will just use those props instead (hence why best to not introduce another way). I need to take a closer look - will reply later in detail.

**Comment by chriskmnds** [comment]
Thanks for pulling this through! I'll give it a more thorough review in a bit, but looks like you nailed the core of it pretty nicely  👍 

I'll drop some suggestions later to wrap this up from my end at least :-)

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-selector.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-selector.tsx]: I think the name here should be `showPlanTypeSelectorDropdown`

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx]: Is this inline function necessary? 🤔

File: client/my-sites/plans-grid/components/plan-type-selector/style.scss
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/style.scss]: Given how we're approaching this, this class name shouldn't be referenced here. Nothing `plans-features-main...` should exist in `plans-grid` stylesheets, as these are consuming end concerns right now. 

I think we will actually going to need this sticky behavior outside of Calypso though, so just moving them to `plans-features-main` may not be ideal. I think we may need to approach this just a tad differently. The code is fine, just the design. How about:

1. Migrate the wrapping with `StickyContainer` into the `plan-type-selector` component.
2. Pass an **optional** property from `plans-features-main` whether we want a sticky component or not e.g. `isSticky` (so defaults to falsy).
3. Rename `plans-features-main__plan-type-selector-sticky-container` to `plan-type-selector-sticky-container`
 
So the end could be something like:

In **plans-features-main** (no wrapping with StickyContainer):

```
<PlanTypeSelector isSticky={ showPlanSelectorDropdown && ! isComparisonGridPlanTypeSelectorInView 
 } />
```

In **plan-type-selector**:

```
const Selector = props.showPlanSelectorDropdown ? (
        <IntervalTypeDropdown />
    ) : (
        <IntervalTypeToggle />
    );

return isSticky ? (
    <StickyContainer><Selector /></StickyContainer>
	) : (
        <Selector />
    );
```

or:

```
return <StickyContainer className={ [ the same as before, just use isSticky in place of isComparisonGridPlanTypeSelectorInView ] }>
    { props.showPlanSelectorDropdown ? (
            <IntervalTypeDropdown />
        ) : (
            <IntervalTypeToggle />
        );
    }
</StickyContainer>

```

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx]: I don't quite follow why you need this and not something you can't add below in `className`? 🤔

File: client/my-sites/plans-grid/components/sticky-container.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/sticky-container.tsx]: Is `stickyClass` not obsolete now since we are passing `className`?

**Review by chriskmnds — CHANGES_REQUESTED** [review]
Thanks again for addressing the feedback in the other PR. It works functionally, albeit with some minor concerns (see notes & gif below). I left some code suggestions though, which I feel will improve this to also be cleaner and reusable. Let me know what you think - I could always be inaccurate :-)

Concerning functionality:

1. I think the breakpoint fires off a few pixels too early, before the features grid goes into mobile view - which hides the buttons in that case.
2. In the comparison grid, it always hides part of the header e.g. the featured labels are always hidden from view.

(you can see both points in the gif below)

https://github.com/Automattic/wp-calypso/assets/1705499/621ded8a-e5ad-4cef-a158-39804e0e2a71

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-selector.tsx
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-selector.tsx]: Ah thanks for catching this 🙏 Making the change now

_Edit:_ https://github.com/Automattic/wp-calypso/pull/85520/commits/26190e5281736f4a7ea10ddbb2727ab0c40250dc

**Comment by jeyip** [comment]
> I think the breakpoint fires off a few pixels too early, before the features grid goes into mobile view - which hides the buttons in that case.

> In the comparison grid, it always hides part of the header e.g. the featured labels are always hidden from view.

Hmmm...I thought I had addressed the needed offsets to avoid obscured sticky plan headers but perhaps not 🤔 Thanks for noting this. I'll take another look along with the suggested refactors

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx]: The sticky container accepts a render prop function as a child, but we don't use the [provided isStuck property](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/components/sticky-container.tsx#L91). Going to give this a quick pass today with fresh eyes and see if not using the `isStuck` property is a code smell though 🤔

File: client/my-sites/plans-grid/components/plan-type-selector/style.scss
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/style.scss]: > Given how we're approaching this, this class name shouldn't be referenced here. Nothing plans-features-main... should exist in plans-grid stylesheets, as these are consuming end concerns right now.

Ah of course. Makes total sense 👍 

> I think we will actually going to need this sticky behavior outside of Calypso though, so just moving them to plans-features-main may not be ideal. I think we may need to approach this just a tad differently. The code is fine, just the design. How about:

> 1. Migrate the wrapping with StickyContainer into the plan-type-selector component.
> 2. Pass an optional property from plans-features-main whether we want a sticky component or not e.g. isSticky (so defaults to falsy).
> 3. Rename plans-features-main__plan-type-selector-sticky-container to plan-type-selector-sticky-container

Considering this now 🤔

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/style.scss]: > I think we will actually going to need this sticky behavior outside of Calypso though

Gotcha, so: 
- Embed plan type selector sticky container inside the `plans-grid`
- Make the plan type selector sticky behavior a setting for the `plans-grid` that consumers can enable or disable

One thing I'll note is that there's the possibility that the plan type selector dropdown will not perform well with longitudinal data analysis. I bring up this consideration because of what happens if we, as Martech, decide to revert back to the original toggle and roll back the mobile sticky behavior. It's also why I originally favored keeping logic outside of the `plans-grid`.

Giving it more thought though, with your design, we have flexibility and can simply disable sticky behavior in that case. The chance at this vestigial code seems like a worthwhile tradeoff for your design that's more resilient to change.

In reality, I doubt we'll roll back the sticky behavior ( it seems like an obvious win ), but I've been wrong plenty in the past 😆

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx]: The stickyClass is only applied when the the child is actually stickied to the screen

File: client/my-sites/plans-grid/components/plan-type-selector/style.scss
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/style.scss]: I started implementing this, but ran into walls 😬. Something about nesting the StickyContainer more deeply within the DOM hierarchy of the plans grid causes the container to revoke "sticky" status way too early. I left a writeup with @southp, who said that he'd continue shepherding this PR while Dehan and I are on break 😎

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/style.scss]: Figured out the issue. @jdc91 I did some digging here and sticky positioning [is always relative to its parent](https://www.w3.org/TR/css-position-3/#sticky-pos). When we move the `StickyContainer` from `PlansFeaturesMain` into the `IntervalTypeSelector` component, there are several wrapping components in between, with [PlanTypeSelector](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/components/plan-type-selector/index.tsx#L20-L22) taking on the new role of "parent".

Because of this, sticky behavior no longer starts and ends when expected. Instead, it fires off and ends almost immediately when scrolling past the PlanTypeSelector container.

Mulling over how to implement Christos's changes effectively 🤔

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/style.scss]: > I think we will actually going to need this sticky behavior outside of Calypso though, so just moving them to plans-features-main may not be ideal.

@chriskmnds Dehan has been asked to implement this feature in landpack, so this use case is now confirmed 👍

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [inline client/my-sites/plans-features-main/index.tsx:772]: **Note:**

I'm changing the DOM structure to have the `StickyContainer` be the direct descendent of the appropriate HTML element in the DOM. This is because of the requirements of `sticky` styling. https://github.com/Automattic/wp-calypso/pull/85520#discussion_r1439866924

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [inline client/my-sites/plans-features-main/index.tsx:703]: **Note:**
TL;DR I'm using javascript instead of CSS to handle logic for mobile screens here for two reasons:
1. it noticeably simplifies logic
2. we're outside of the `plans-grid` shared package so it doesn't clash with any conventions ( which rely on CSS, and not JS, for mobile logic )

___

We want sticky behavior _only_ for mobile screens.

We chatted about, where possible, keeping desktop vs. tablet vs. mobile logic to CSS media queries in the `plans-grid` for the sake of convention. So why the switch to Javascript?

I do this because relying on media queries for our use case ends up making feature logic more complex. We would effectively replicate behavior that the `StickyContainer` already supports through programmatic javascript: [the ability to enable and disable sticky behavior under different conditions]( https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/components/comparison-grid/index.tsx#L1094 ). To avoid issues with existing conventions, I move mobile "definitions" into `plans-features-main` by using `isMobile` from the `'@automattic/viewport'` package.

File: client/my-sites/plans-features-main/index.tsx
Comment (ddc22) [reply client/my-sites/plans-features-main/index.tsx:772]: If this is the only reason you changed the dom structure and their is no other intrinsic code hygiene value I would suggest to change the dom structure back and just change the position attribute to fixed.
link: https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block#identifying_the_containing_block

Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx:772]: Responded here p1704303068235729/1704249392.584319-slack-CV2TX2PAN and here p1704303465678879/1704249392.584319-slack-CV2TX2PAN

> Hmmm…not sure about the structure of the PR then. If that’s the case, then we’d probably want to forgo reusing the sticky container entirely, which forcibly assigns a sticky css attribute

> Also, I might be missing something, but how do we detect when the user has scrolled to a certain point to trigger the fixed positioning? :thinking_face: It seems like it would make our logic inherently more complex

File: client/my-sites/plans-grid/components/sticky-container.tsx
Comment (ddc22) [reply client/my-sites/plans-grid/components/sticky-container.tsx]: @jeyip maybe you missed addressing this comment?

File: client/my-sites/plans-grid/components/plan-type-selector/style.scss
Comment (ddc22) [inline client/my-sites/plans-grid/components/plan-type-selector/style.scss]: Cleanup required probably?

**Review by ddc22 — COMMENTED** [review]
LGTM 👍 
I have left a few minor comments. This can be shipped once resolved :D

File: client/my-sites/plans-grid/components/plan-type-selector/style.scss
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/style.scss]: Thank you! I double checked this and it looks fine since there are `calypso` scss path aliases used throughout the `plans-grid` package.

Removing the inline comment now [95849e9](https://github.com/Automattic/wp-calypso/pull/85520/commits/95849e904bca9a67aac070512b088fecdccdec40)

File: client/my-sites/plans-grid/components/sticky-container.tsx
Comment (jeyip) [reply client/my-sites/plans-grid/components/sticky-container.tsx]: Thanks for the ping @jdc91! I could've been clearer here. This comment is addressed in https://github.com/Automattic/wp-calypso/pull/85520#discussion_r1435439562, but I may not have elaborated well 😅

`stickyClass` is only applied to the component if it is actually "stuck" to the screen. `className` was meant to allow the consumer of the `StickyContainer` to apply a `className` that would be present on the container, regardless of its "sticky" state. After the revisions made here https://github.com/Automattic/wp-calypso/pull/85520#discussion_r1440110172, however, and the shift away from using media queries, `className` itself is no longer needed. I'll go ahead and remove it for the time being

**Comment by jeyip** [comment]
I see failing e2e specs across the board, so I'll wait until tomorrow morning to merge this. @chriskmnds when you're available, if you have any concerns at all, feel free to drop a message and I'll be happy to return and address any issues

**Pull request merged**
